plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/graphvisitors/InteractiveLayeredGraphVisitor.java

```diff
@@ -0,0 +1,451 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0  *******************************************************************************/
+package org.eclipse.elk.alg.layered.graphvisitors;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.elk.alg.layered.options.CycleBreakingStrategy;
+import org.eclipse.elk.alg.layered.options.LayerConstraint;
+import org.eclipse.elk.alg.layered.options.LayeredOptions;
+import org.eclipse.elk.alg.layered.options.LayeringStrategy;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.options.CoreOptions;
+import org.eclipse.elk.core.options.Direction;
+import org.eclipse.elk.core.util.IGraphElementVisitor;
+import org.eclipse.elk.graph.ElkConnectableShape;
+import org.eclipse.elk.graph.ElkEdge;
+import org.eclipse.elk.graph.ElkGraphElement;
+import org.eclipse.elk.graph.ElkNode;
+import org.eclipse.elk.graph.ElkPort;
+
+/**
+ * Graph visitor which visits only the root node and recursively steps through the graph
+ * to set interactive option for the layered algorithm.
+ */
+public class InteractiveLayeredGraphVisitor implements IGraphElementVisitor {
+
+    /**
+     * Constant help with node placement, since nodes should not overlap and the their width and height may not be
+     * final.
+     */
+    public static final int NODE_PLACEMENT_HELPER = 100000;
+    
+    /**
+     * Set as the layer of a node that should be in the last layer.
+     */
+    public static final int LAST_LAYER = 100000;
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see org.eclipse.elk.core.util.IGraphElementVisitor#visit(org.eclipse.elk.graph.ElkGraphElement)
+     */
+    @Override
+    public void visit(final ElkGraphElement element) {
+        // Only apply to root of the graph
+        if (element.eContainer() == null && element instanceof ElkNode) {
+            ElkNode root = (ElkNode) element;
+            setInteractiveOptions(root);
+        }
+    }
+
+    /**
+     * Sets the coordinates of the nodes in the graph.
+     * 
+     * @param root
+     *            Root of the graph
+     */
+    private void setInteractiveOptions(final ElkNode root) {
+        String algorithm = root.getProperty(CoreOptions.ALGORITHM);
+        if (algorithm == null || LayeredOptions.ALGORITHM_ID.endsWith(algorithm)) {
```

Comment:
This code relies on the fact that ELK Layered is currently the default algorithm. That should at least be mentioned in the comments somewhere, if not changed. I'm actually not sure whether `algorithm` can be `null` at this point. There is an `IGraphElementVisitor` which explicitly sets layout algorithms on each node.

Reply:
`algorithm` can be null at this point.